### PR TITLE
Add collect list to dask-cudf groupby aggregations

### LINKED ISSUE #7812: [BUG] Groupby collect list fails with Dask
cuDF recently implemented groupby collect list. I'd like to be able to use this with Dask, like I can on the CPU. Currently, it looks like Dask's collect list goes a codepath that requires iterating through the object, which we explicitly don't permit (https://github.com/rapidsai/cudf/issues/7481). We may want to explore this in Dask or special case in Dask cuDF.

```python-traceback
import cudf
import dask_cudf
import pandas as pd
import dask.dataframe as dd
​
from io import StringIO

​
data = """a,b
1595802,1611:0.92
1595802,1610:0.07
1524246,1807:0.92
1524246,1608:0.07"""
​
df = pd.read_csv(StringIO(data))
ddf = dd.from_pandas(df, 2)
​
gdf = cudf.from_pandas(df)
gddf = dask_cudf.from_cudf(gdf, 2)
​
print(ddf.groupby("a").agg({"b":list}).compute()) # works as expected
print(gddf.groupby("a").agg({"b":list}).compute())
                              b
a                              
1595802  [1611:0.92, 1610:0.07]
1524246  [1807:0.92, 1608:0.07]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname, udf)
    179     try:
--> 180         yield
    181     except Exception as e:

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   5507     with raise_on_meta_error(funcname(func), udf=kwargs.pop("udf", False)):
-> 5508         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   5509 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py in _groupby_apply_funcs(df, *index, **kwargs)
    920     for result_column, func, func_kwargs in funcs:
--> 921         r = func(grouped, **func_kwargs)
    922 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py in _apply_func_to_column(df_like, column, func)
    966 
--> 967     return func(df_like[column])
    968 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py in <lambda>(s)
    843                 _apply_func_to_column,
--> 844                 dict(column=input_column, func=lambda s: s.apply(list)),
    845             )

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/core/groupby/groupby.py in apply(self, function)
    422         ]
--> 423         chunk_results = [function(chk) for chk in chunks]
    424 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/core/groupby/groupby.py in <listcomp>(.0)
    422         ]
--> 423         chunk_results = [function(chk) for chk in chunks]
    424 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/core/series.py in __iter__(self)
   1189     def __iter__(self):
-> 1190         cudf.utils.utils.raise_iteration_error(obj=self)
   1191 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/utils/utils.py in raise_iteration_error(obj)
    358     raise TypeError(
--> 359         f"{obj.__class__.__name__} object is not iterable. "
    360         f"Consider using `.to_arrow()`, `.to_pandas()` or `.values_host` "

TypeError: Series object is not iterable. Consider using `.to_arrow()`, `.to_pandas()` or `.values_host` if you wish to iterate over the values.

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-16-08ab17754440> in <module>
     22 
     23 print(ddf.groupby("a").agg({"b":list}).compute()) # works as expected
---> 24 print(gddf.groupby("a").agg({"b":list}).compute())

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py in agg(self, arg, split_every, split_out)
   1846     @derived_from(pd.core.groupby.DataFrameGroupBy)
   1847     def agg(self, arg, split_every=None, split_out=1):
-> 1848         return self.aggregate(arg, split_every=split_every, split_out=split_out)
   1849 
   1850 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask_cudf/groupby.py in aggregate(self, arg, split_every, split_out)
     81 
     82         return super().aggregate(
---> 83             arg, split_every=split_every, split_out=split_out
     84         )
     85 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py in aggregate(self, arg, split_every, split_out)
   1842             return self.size()
   1843 
-> 1844         return super().aggregate(arg, split_every=split_every, split_out=split_out)
   1845 
   1846     @derived_from(pd.core.groupby.DataFrameGroupBy)

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py in aggregate(self, arg, split_every, split_out)
   1622             split_out=split_out,
   1623             split_out_setup=split_out_on_index,
-> 1624             sort=self.sort,
   1625         )
   1626 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/core.py in apply_concat_apply(args, chunk, aggregate, combine, meta, token, chunk_kwargs, aggregate_kwargs, combine_kwargs, split_every, split_out, split_out_setup, split_out_setup_kwargs, sort, ignore_index, **kwargs)
   5459 
   5460     if meta is no_default:
-> 5461         meta_chunk = _emulate(chunk, *args, udf=True, **chunk_kwargs)
   5462         meta = _emulate(
   5463             aggregate, _concat([meta_chunk], ignore_index), udf=True, **aggregate_kwargs

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   5506     """
   5507     with raise_on_meta_error(funcname(func), udf=kwargs.pop("udf", False)):
-> 5508         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   5509 
   5510 

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/contextlib.py in __exit__(self, type, value, traceback)
    128                 value = type()
    129             try:
--> 130                 self.gen.throw(type, value, traceback)
    131             except StopIteration as exc:
    132                 # Suppress StopIteration *unless* it's the same exception that

/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname, udf)
    199         )
    200         msg = msg.format(" in `{0}`".format(funcname) if funcname else "", repr(e), tb)
--> 201         raise ValueError(msg) from e
    202 
    203 

ValueError: Metadata inference failed in `_groupby_apply_funcs`.

You have supplied a custom function and Dask is unable to 
determine the type of output that that function returns. 

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
TypeError('Series object is not iterable. Consider using `.to_arrow()`, `.to_pandas()` or `.values_host` if you wish to iterate over the values.')

Traceback:
---------
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/utils.py", line 180, in raise_on_meta_error
    yield
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/core.py", line 5508, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 921, in _groupby_apply_funcs
    r = func(grouped, **func_kwargs)
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 967, in _apply_func_to_column
    return func(df_like[column])
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 844, in <lambda>
    dict(column=input_column, func=lambda s: s.apply(list)),
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/core/groupby/groupby.py", line 423, in apply
    chunk_results = [function(chk) for chk in chunks]
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/core/groupby/groupby.py", line 423, in <listcomp>
    chunk_results = [function(chk) for chk in chunks]
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/core/series.py", line 1190, in __iter__
    cudf.utils.utils.raise_iteration_error(obj=self)
  File "/raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331/lib/python3.7/site-packages/cudf/utils/utils.py", line 359, in raise_iteration_error
    f"{obj.__class__.__name__} object is not iterable. "
```

```
!conda list | grep "rapids\|dask\|numpy\|cupy\|arrow\|pandas"
# packages in environment at /raid/nicholasb/miniconda3/envs/rapids-gpubdb-20210331:
arrow-cpp                 1.0.1           py37h2318771_14_cuda    conda-forge
arrow-cpp-proc            3.0.0                      cuda    conda-forge
cudf                      0.19.0a210331   cuda_10.2_py37_gc99fcef41b_313    rapidsai-nightly
cuml                      0.19.0a210331   cuda10.2_py37_g83168076c_138    rapidsai-nightly
cupy                      8.6.0            py37h7fc54ca_0    conda-forge
dask                      2021.3.1           pyhd8ed1ab_0    conda-forge
dask-core                 2021.3.1           pyhd8ed1ab_0    conda-forge
dask-cuda                 0.19.0a210331           py37_45    rapidsai-nightly
dask-cudf                 0.19.0a210331   py37_gc99fcef41b_313    rapidsai-nightly
libcudf                   0.19.0a210331   cuda10.2_gbe2f0c000f_314    rapidsai-nightly
libcuml                   0.19.0a210331   cuda10.2_g83168076c_138    rapidsai-nightly
libcumlprims              0.19.0a210316   cuda10.2_ge7e82a0_12    rapidsai-nightly
librmm                    0.19.0a210331   cuda10.2_g9d1ba02_50    rapidsai-nightly
numpy                     1.19.5           py37haa41c4c_1    conda-forge
pandas                    1.2.3            py37hdc94413_0    conda-forge
pyarrow                   1.0.1           py37hbeecfa9_14_cuda    conda-forge
rmm                       0.19.0a210331   cuda_10.2_py37_g9d1ba02_50    rapidsai-nightly
ucx                       1.9.0+gcd9efd3       cuda10.2_0    rapidsai-nightly
ucx-proc                  1.0.0                       gpu    rapidsai-nightly
ucx-py                    0.19.0a210331   py37_gcd9efd3_46    rapidsai-nightly
```

### PR DESCRIPTION
**Pull request by charlesbluca**
Closes #7812

Adds support for cuDF's `collect` aggregation in dask-cuDF.

### CODE REVIEW

**Comment by charlesbluca** [comment]
To illustrate the problem:

```python
>>> data = """a,b
    1595802,1611:0.92
    1595802,1610:0.07
    1524246,1807:0.92
    1524246,1608:0.07"""

>>> df = pd.read_csv(StringIO(data))
>>> ddf = dd.from_pandas(df, 2)

>>> gdf = cudf.from_pandas(df)
>>> gddf = dask_cudf.from_cudf(gdf, 2)

>>> print(ddf.groupby("a").agg({"b":list}).compute())
                              b
a                              
1595802  [1611:0.92, 1610:0.07]
1524246  [1807:0.92, 1608:0.07]
>>> print(gddf.groupby("a").agg({"b":list}).compute())
                                b
a                                
1524246  [[1807:0.92, 1608:0.07]]
1595802  [[1611:0.92, 1610:0.07]]
```

File: python/dask_cudf/dask_cudf/groupby.py [hunk 1]
```diff
@@ -255,7 +255,12 @@ def groupby_agg(
         # be str, rather than tuples).
         for col in aggs:
             _aggs[col] = _aggs[col][0]
-    _meta = ddf._meta.groupby(gb_cols, as_index=as_index).agg(_aggs)
+    try:
+        _meta = ddf._meta.groupby(gb_cols, as_index=as_index).agg(_aggs)
+    except NotImplementedError:
+        _meta = ddf._meta_nonempty.groupby(gb_cols, as_index=as_index).agg(
+            _aggs
+        )
```
Comment (kkraus14) [inline python/dask_cudf/dask_cudf/groupby.py:263]: Did the `list` aggregation not work on an empty dataframe here or return a different type than expected or something?

Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/groupby.py:263]: This also occurred on non-empty dataframes; here's the traceback:

```
Traceback (most recent call last):
  File "cudf/_lib/groupby.pyx", line 164, in cudf._lib.groupby.GroupBy.aggregate
RuntimeError: cuDF failure at: /raid/charlesb/github/cudf/cpp/src/column/column_factories.cpp:68: make_empty_column is invalid to call on nested types

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "python/dask_cudf/test_dask_collect.py", line 21, in <module>
    print(gddf.groupby("a").agg({"b":list}).compute())
  File "/raid/charlesb/github/dask/dask/dataframe/groupby.py", line 1847, in agg
    return self.aggregate(arg, split_every=split_every, split_out=split_out)
  File "/raid/charlesb/github/cudf/python/dask_cudf/dask_cudf/groupby.py", line 70, in aggregate
    return groupby_agg(
  File "/raid/charlesb/github/cudf/python/dask_cudf/dask_cudf/groupby.py", line 259, in groupby_agg
    _meta = ddf._meta.groupby(gb_cols, as_index=as_index).agg(_aggs)
  File "/raid/charlesb/miniconda3/envs/cudf_dev/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/raid/charlesb/miniconda3/envs/cudf_dev/lib/python3.8/site-packages/cudf-0.20.0a0+170.gb0218b45e7.dirty-py3.8-linux-x86_64.egg/cudf/core/groupby/groupby.py", line 164, in agg
    result = self._groupby.aggregate(self.obj, normalized_aggs)
  File "cudf/_lib/groupby.pyx", line 173, in cudf._lib.groupby.GroupBy.aggregate
NotImplementedError: Aggregation not supported for empty columns
```

Which was happening when trying to do:

https://github.com/rapidsai/cudf/blob/32222de53b6f97fb26f01ee4e413c72019c0102c/python/cudf/cudf/_lib/groupby.pyx#L171-L175

Comment (kkraus14) [reply python/dask_cudf/dask_cudf/groupby.py:263]: cc @shwina do you happen to know if this is a Python thing or a libcudf thing?

Comment (vyasr) [reply python/dask_cudf/dask_cudf/groupby.py:292]: It seems like the underlying blocker is `libcudf`, specifically #7611. There appears to be some confusion on the nature of the problem in #7611, perhaps @mythrocks can provide some context as to whether this is actually something Python `cudf` needs to be disabling or if the issue raised was erroneous and we would be safe to reenable this in Python.

Comment (mythrocks) [reply python/dask_cudf/dask_cudf/groupby.py:292]: I was hoping to address the rolling-window side of #7611 some time this week, if all goes well.

Do I understand correctly that you're running into this in the group-by case, for empty input?

Comment (vyasr) [reply python/dask_cudf/dask_cudf/groupby.py:292]: Just to close the loop here, Mithun and I chatted offline to understand where the underlying problem is. It's definitely in `libcudf`, not `cudf` Python. Fortunately it looks like `dask_cudf` is capable of handling this for the moment, so hopefully it's not a blocker.

Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/groupby.py:292]: Just for clarification, this case happens for non-empty input - in particular, the example above:

```python
data = """a,b
    1595802,1611:0.92
    1595802,1610:0.07
    1524246,1807:0.92
    1524246,1608:0.07"""

df = pd.read_csv(StringIO(data))
ddf = dd.from_pandas(df, 2)

gdf = cudf.from_pandas(df)
gddf = dask_cudf.from_cudf(gdf, 2)

print(ddf.groupby("a").agg({"b":list}).compute())
print(gddf.groupby("a").agg({"b":list}).compute())
```

Comment (mythrocks) [reply python/dask_cudf/dask_cudf/groupby.py:292]: Hello, @charlesbluca. Would it be viable to retest your case with the fix I have up on #8279?

Comment (mythrocks) [reply python/dask_cudf/dask_cudf/groupby.py:292]: P.S. Sorry for the wait.

Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/groupby.py:292]: Can verify that #8279 resolves this! Thanks @mythrocks 😃

File: python/dask_cudf/dask_cudf/tests/test_groupby.py [hunk 2]
```diff
@@ -125,6 +125,38 @@ def test_groupby_std(func):
...
+
+    ddf = dask_cudf.from_cudf(gdf, npartitions=5)
+
+    a = func(gdf).to_pandas()
+    b = func(ddf).compute().to_pandas()
+
+    a.index.name = None
+    a.name = None
+    b.index.name = None
+    b.name = None
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/tests/test_groupby.py:155]: This concerns me a bit - Do the index names not match for "collect"?

File: python/dask_cudf/dask_cudf/tests/test_groupby.py [hunk 3]
```diff
@@ -125,6 +125,38 @@ def test_groupby_std(func):
...
 
+@pytest.mark.parametrize(
+    "func",
+    [
+        lambda df: df.groupby("x").agg({"y": "collect"}),
+        pytest.param(
+            lambda df: df.groupby("x").y.agg("collect"), marks=pytest.mark.skip
+        ),
+    ],
+)
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/tests/test_groupby.py:136]: Is there any reason to define `func` this way?  Am I misunderstanding, or will the second of two cases always skipped?

**Review by rjzamora — COMMENTED** [review]
This looks great @charlesbluca - Thanks!

My only concern is that you are explicity setting the index names to `None` in the test.  Is the "collection" result somehow different from other aggregations?

File: python/dask_cudf/dask_cudf/tests/test_groupby.py (hunk 3 above)
Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/tests/test_groupby.py:136]: This param skip, and the index nulling, I lifted from another dask-cudf groupby test:

https://github.com/rapidsai/cudf/blob/5b8895dc50a5daf99bafede67617e8ebd39d3af6/python/dask_cudf/dask_cudf/tests/test_groupby.py#L47-L77

I can see what happens when we don't set the index to `None` here, but when that test isn't skipped we get an `AssertionError` on the types:

```python
__________________________________________ test_groupby_collect[<lambda>1] ___________________________________________

func = <function <lambda> at 0x7f3c804a0d30>

    @pytest.mark.parametrize(
        "func",
        [
            lambda df: df.groupby("x").agg({"y": "collect"}),
            lambda df: df.groupby("x").y.agg("collect"),
        ],
    )
    def test_groupby_collect(func):
        pdf = pd.DataFrame(
            {
                "x": np.random.randint(0, 5, size=10000),
                "y": np.random.normal(size=10000),
            }
        )
    
        gdf = cudf.DataFrame.from_pandas(pdf)
    
        ddf = dask_cudf.from_cudf(gdf, npartitions=5)
    
        a = func(gdf).to_pandas()
        b = func(ddf).compute().to_pandas()
    
        a.index.name = None
        a.name = None
        b.index.name = None
        b.name = None
    
>       dd.assert_eq(a, b)

dask_cudf/tests/test_groupby.py:155: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../compose/etc/conda/cuda_11.2.72/envs/rapids/lib/python3.8/site-packages/dask/dataframe/utils.py:559: in assert_eq
    tm.assert_series_equal(a, b, check_names=check_names, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = 0    [-0.02279966962796973, -0.2268040371246616, 0....
1    [1.0547561143327269, 0.07632651478542447, -0.0...
2    [1.... [0.35305396010499146, 2.022936601816015, -0.02...
4    [0.7639835327097312, 0.9458744987601149, 0.370...
dtype: object
right =                                                    y
0  [-0.02279966962796973, -0.2268040371246616, 0....
1  [1.054756...796, 1.498...
3  [0.35305396010499146, 2.022936601816015, -0.02...
4  [0.7639835327097312, 0.9458744987601149, 0.370...
cls = <class 'pandas.core.series.Series'>

    def _check_isinstance(left, right, cls):
        """
        Helper method for our assert_* methods that ensures that
        the two objects being compared have the right type before
        proceeding with the comparison.
    
        Parameters
        ----------
        left : The first object being compared.
        right : The second object being compared.
        cls : The class type to check against.
    
        Raises
        ------
        AssertionError : Either `left` or `right` is not an instance of `cls`.
        """
        cls_name = cls.__name__
    
        if not isinstance(left, cls):
            raise AssertionError(
                f"{cls_name} Expected type {cls}, found {type(left)} instead"
            )
        if not isinstance(right, cls):
>           raise AssertionError(
                f"{cls_name} Expected type {cls}, found {type(right)} instead"
            )
E           AssertionError: Series Expected type <class 'pandas.core.series.Series'>, found <class 'pandas.core.frame.DataFrame'> instead
```

It looks like we are creating a dataframe here when we should be making a series.

File: python/dask_cudf/dask_cudf/tests/test_groupby.py (hunk 2 above)
Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/tests/test_groupby.py:155]: I copied this over from another test, but it looks like it isn't needed - I can remove these lines.

File: python/dask_cudf/dask_cudf/tests/test_groupby.py (hunk 3 above)
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/tests/test_groupby.py:136]: >It looks like we are creating a dataframe here when we should be making a series.

Ah - It seems like this was already a problem before this PR.  In that case, it is probably okay to fix that in a follow-up PR.

Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/tests/test_groupby.py:136]: Sure, do you know if there is an open issue for this problem? If not, I can open one so we can keep track of the follow up fix.

**Review by rjzamora — APPROVED** [review]
Thanks again @charlesbluca - Note that I filed #8655 to make sure we address the Series/DataFrame inconsistency discussed here.

**Comment by charlesbluca** [comment]
Thanks for opening the issue @rjzamora!

**Comment by charlesbluca** [comment]
@gpucibot merge

**Pull request merged**